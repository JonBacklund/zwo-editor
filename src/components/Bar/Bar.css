.segment {
  position: relative;
}
.bar {    
  border-radius: 10px;
  border: 1px solid white;
}

.label {
  position: absolute;
  top: -40px;
  left: 0;
  color: white;
  background-color: rgba(0, 0, 0, 0.7);
  width: 110px;
  border-radius: 5px;
  margin: 5px 0;
  text-align: center;
  font-size: 14px;
  overflow: hidden;
  z-index: 100;
}

.comment {
  position: absolute;
  top: -60px;
  left: 105px;
  color: gray;
  padding: 5px;  
  font-size: 20px;
}