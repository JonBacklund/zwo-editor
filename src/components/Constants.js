export const Colors = {
  GRAY: '#807F80',
  BLUE: '#0E90D4',
  GREEN: '#00C46A',
  YELLOW: '#FFCB00',
  ORANGE: '#FF6430',
  RED: '#E90000'
}