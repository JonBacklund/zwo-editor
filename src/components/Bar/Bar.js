import React, { useState } from 'react'
import './Bar.css'
import { Colors, Zones } from '../Constants'
import { Resizable } from 're-resizable'
import moment from 'moment'
import 'moment-duration-format'
import Label from '../Label/Label'

const Bar = ({ id, time, power, ftp, onChange, onClick }) => {

  const multiplier = 250
  const timeMultiplier = 5

  const powerLabel = Math.round(power * ftp)
  const durationLabel = getDuration(time)
  const style = zwiftStyle(power)

  const [width, setWidth] = useState(time / timeMultiplier)
  const [height, setHeight] = useState(power * multiplier)

  const [showLabel, setShowLabel] = useState(false)

  const handleResizeStop = ({ e, direction, ref, d }) => {
    setWidth(width + d.width)
    setHeight(height + d.height)
    onChange(id, { time: (width + d.width) * timeMultiplier, power: (height + d.height) / multiplier, type: 'bar', id: id })
  }

  const handleResize = ({ e, direction, ref, d }) => {
    onChange(id, { time: (width + d.width) * timeMultiplier, power: (height + d.height) / multiplier, type: 'bar', id: id })
  }

  function getDuration(seconds) {
    // 1 pixel equals 5 seconds 
    return moment.duration(seconds, "seconds").format("mm:ss", { trim: false })
  }

  function zwiftStyle(zone) {

    if (zone >= 0 && zone < Zones.Z1.max) {
      // Z1 gray
      return { backgroundColor: Colors.GRAY }
    } else if (zone >= Zones.Z2.min && zone < Zones.Z2.max) {
      // Z2 blue
      return { backgroundColor: Colors.BLUE }
    } else if (zone >= Zones.Z3.min && zone < Zones.Z3.max) {
      // Z3 green
      return { backgroundColor: Colors.GREEN }
    } else if (zone >= Zones.Z4.min && zone < Zones.Z4.max) {
      // Z4 yellow
      return { backgroundColor: Colors.YELLOW }
    } else if (zone >= Zones.Z5.min && zone < Zones.Z5.max) {
      // Z5 orange      
      return { backgroundColor: Colors.ORANGE }
    } else {
      // Z6 red          
      return { backgroundColor: Colors.RED }
    }
  }

  return (
    <div className='segment'
      onMouseEnter={() => setShowLabel(true)}
      onMouseLeave={() => setShowLabel(false)}
    >
      {showLabel &&
        <Label duration={durationLabel} power={powerLabel} />
      }      
      <Resizable
        className='bar'
        size={{
          width: width,
          height: height,
        }}
        minWidth={3}
        minHeight={multiplier * Zones.Z1.min}
        maxHeight={multiplier * Zones.Z6.max}
        enable={{ top: true, right: true }}
        grid={[1, 1]}
        onResizeStop={(e, direction, ref, d) => handleResizeStop({ e, direction, ref, d })}
        onResize={(e, direction, ref, d) => handleResize({ e, direction, ref, d })}
        onClick={() => onClick(id)}
        style={style}
      >
      </Resizable>
    </div>

  );
}

export default Bar