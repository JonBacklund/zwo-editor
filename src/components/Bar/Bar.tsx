import React, { useState, useEffect } from 'react'
import './Bar.css'
import { Colors, Zones } from '../Constants'
import { Resizable } from 're-resizable'
import moment from 'moment'
import 'moment-duration-format'
import Label from '../Label/Label'

function round5(x: number)
{
  return Math.ceil(x/5)*5;
}

const Bar = (props: { id: string, time: number, power: number, cadence: number, ftp: number, weight: number, onChange: Function, onClick: Function, selected: boolean }) => {

  const multiplier = 250
  const timeMultiplier = 3

  const powerLabel = Math.round(props.power * props.ftp)
  const durationLabel = getDuration(props.time)
  const style = zwiftStyle(props.power)

  const [width, setWidth] = useState(props.time / timeMultiplier)
  const [height, setHeight] = useState(props.power * multiplier)  

  const [showLabel, setShowLabel] = useState(false)

  const [selected, setSelected] = useState(props.selected)
  useEffect(()=>{
    setSelected(props.selected)
  },[props.selected])

  const handleResizeStop = (dWidth: number, dHeight: number) => {
    setWidth(width + dWidth)
    setHeight(height + dHeight)
    props.onChange(props.id, { time: round5((width + dWidth) * timeMultiplier), power: (height + dHeight) / multiplier, cadence: props.cadence, type: 'bar', id: props.id })
  }

  const handleResize = (dWidth: number, dHeight: number) => {
    props.onChange(props.id, { time: round5((width + dWidth) * timeMultiplier), power: (height + dHeight) / multiplier, cadence: props.cadence, type: 'bar', id: props.id })
  }

  function getDuration(seconds: number) {
    // 1 pixel equals 5 seconds 
    return moment.duration(seconds, "seconds").format("mm:ss", { trim: false })
  }

  function zwiftStyle(zone: number) {

    if (zone >= 0 && zone < Zones.Z1.max) {
      // Z1 gray
      return { backgroundColor: Colors.GRAY }
    } else if (zone >= Zones.Z2.min && zone < Zones.Z2.max) {
      // Z2 blue
      return { backgroundColor: Colors.BLUE }
    } else if (zone >= Zones.Z3.min && zone < Zones.Z3.max) {
      // Z3 green
      return { backgroundColor: Colors.GREEN }
    } else if (zone >= Zones.Z4.min && zone < Zones.Z4.max) {
      // Z4 yellow
      return { backgroundColor: Colors.YELLOW }
    } else if (zone >= Zones.Z5.min && zone < Zones.Z5.max) {
      // Z5 orange      
      return { backgroundColor: Colors.ORANGE }
    } else {
      // Z6 red          
      return { backgroundColor: Colors.RED }
    }
  }

  return (
    <div className='segment'
      onMouseEnter={() => setShowLabel(true)}
      onMouseLeave={() => setShowLabel(false)}
      onClick={() => props.onClick(props.id)}
      style={props.selected ? { zIndex: 10 } : {}}
    >
      {(selected || showLabel) &&
        <Label duration={durationLabel} power={powerLabel} weight={props.weight} ftp={props.ftp} />
      }
      <Resizable
        className='bar'
        size={{
          width: props.time / timeMultiplier,
          height: props.power * multiplier,
        }}
        minWidth={3}
        minHeight={multiplier * Zones.Z1.min}
        maxHeight={multiplier * Zones.Z6.max}
        enable={{ top: true, right: true }}
        grid={[1, 1]}
        onResizeStop={(e, direction, ref, d) => handleResizeStop(d.width, d.height)}
        onResize={(e, direction, ref, d) => handleResize(d.width, d.height)}
        style={style}
      >
      </Resizable>
    </div>

  );
}

export default Bar