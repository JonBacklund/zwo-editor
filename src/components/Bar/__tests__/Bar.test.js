import React from 'react';
import Bar from '../../Bar/Bar';
import { Zones } from '../../Constants'
import renderer from 'react-test-renderer';
import {cleanup, fireEvent, render} from '@testing-library/react';
import { v4 as uuidv4 } from 'uuid'
import '@testing-library/jest-dom/extend-expect'


test('Bar renders correctly', () => {

  const bar = {
    time: 50,
    power: Zones.Z3.min,
    type: 'bar',
    id: uuidv4()
  }

  const ftp = 250

  const component = renderer.create(    
      <Bar
        key={bar.id}
        id={bar.id}
        time={bar.time}
        power={bar.power}
        ftp={ftp}
        onChange={() => handleOnChange}
        onClick={() => handleOnClick}
      />
  )

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
})