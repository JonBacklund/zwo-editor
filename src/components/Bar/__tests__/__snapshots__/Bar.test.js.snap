// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bar renders correctly 1`] = `
<div
  className="segment"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="bar"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#00C46A",
        "boxSizing": "border-box",
        "flexShrink": 0,
        "height": "190.75px",
        "maxHeight": 500,
        "maxWidth": undefined,
        "minHeight": 25,
        "minWidth": 3,
        "position": "relative",
        "userSelect": "auto",
        "width": "10px",
      }
    }
  >
    <span>
      <div
        className=""
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "cursor": "row-resize",
            "height": "10px",
            "left": "0px",
            "position": "absolute",
            "top": "-5px",
            "userSelect": "none",
            "width": "100%",
          }
        }
      />
      <div
        className=""
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "cursor": "col-resize",
            "height": "100%",
            "position": "absolute",
            "right": "-5px",
            "top": "0px",
            "userSelect": "none",
            "width": "10px",
          }
        }
      />
    </span>
  </div>
</div>
`;
